src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        this.head = null;
+        this.tail = null;
+        this.size = 0;
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            Node<T> newNode = new Node<>(null, value, null);
+            head = newNode;
+            tail = newNode;
+        } else {
+            Node<T> newNode = new Node<>(head, value, null);
+            head.next = newNode;
+            head = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index != size) {
+            checkIndex(index);
+        } else if (isEmpty() || index == size) {
+            add(value);
+            return;
+        }
+        Node<T> current = getNode(index);
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        if (current.prev != null) {
+            current.prev.next = newNode;
+        } else {
+            tail = newNode;
+        }
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNode(index);
+        T oldItem = current.item;
+        current.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = tail;
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else if (index == 0) {
+            tail = current.next;
+            tail.prev = null;
+        } else if (index == size - 1) {
+            current = head;
+            head = current.prev;
+            head.next = null;
+        } else {
+            current = getNode(index);
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        T oldItem = current.item;
+        size--;
+        return oldItem;
     }
 
     @Override
     public boolean remove(T object) {
+        int index = getNode(object);
+        if (index >= 0) {
+            remove(index);
+            return true;
+        }
```

Comment:
now you complicate your solution and can iterate by list twice. One time is enough